Resolve issues with limited file matches in large repositories

itsbjoern opened the pull request:
First off, this is a stacked PR based on PR https://github.com/zigcBenx/storm-search/pull/22, a lot of those changes should probably be reviewed in there first.

This PR aims to close issue #20 by addressing the previous search limits. In the previous engine the search was having to churn through all file results at once before responding with any results which made it really slow for large result sets.

My changes address this in the following way:
1. Search files are scanned for the query in batches, and then also sent to the webview in batches once processed. This way we can avoid choking up the UI completely for really large result sets (or really short queries)
2. The webview only renders a partial results until scrolled to the end of the list to not block the UI. With thousands of matches having to render it can get really slow, real fast. It still always keeps updating the total found so far
3. There is a cache for file matches in the webview, this is cleared now when a document changes to ensure the most up to date version is found

I snuck in a change which allows multiple matches within the same line to be highlighted correctly (and show up as separate matches). I think this was planned already as "column" already existed as a field.

zigcBenx commented on src/WebviewManager.ts:
```diff
@@ -106,14 +125,37 @@ export class WebviewManager {
+                });
+            }
+
+            if (resultCount === 0) {
```
I think I managed to reproduce your assumption.

I will really appreciate it if you manage to fix this :pray:

itsbjoern commented:
Hey @zigcBenx I think I managed to fix the last problems here.

1. The manager now returns early instead of just breaking when the term changes, so the noResults message isn't sent in that instance https://github.com/zigcBenx/storm-search/pull/23/files#diff-5764a726641157681697107b95b774c54e73ea3e654a3fc5556d3b4708d6a471R139
2. I forgot to move up the code that handles updating the match and file count even if the overall result rendering is halted. I fixed that now.
3. There was a mismatch of the highlight between the preview snippet and the preview location for multi-column matches. I think i now managed to figure that out. I did have to limit the preview (50 chars before and after the match) as this was causing trouble for some files. E.g., for sourcemaps etc where hundreds or thousands of characters can appear in one line. I fixed this by separately tracking the highlight location for the snippet and the actual code itself.

There is quite a few things that would still be nice to have (for example a stable file order), but I'll refrain from lumping them into this PR.

itsbjoern commented:
Here's an example, even if copying the whole node_modules and getting thousands of file matches and hundreds of thousands of code matches it still performs pretty damn fast.

https://github.com/user-attachments/assets/6f194e02-d836-421d-a15d-d231cd352ca9

zigcBenx commented:
When I compile your version I get this error:
```
src/WebviewManager.ts:122:86 - error TS2345: Argument of type 'number | undefined' is not assignable to parameter of type 'number'.
  Type 'undefined' is not assignable to type 'number'.

122                     await this.handleOpenFile(panel, message.filePath, message.line, message.column);
                                                                                         ~~~~~~~~~~~~~~


Found 1 error in src/WebviewManager.ts:122
```

Could you fix this?
Maybe we coul use `message.column ?? 0` at the call site, or change `handleOpenFile` to accept `column?: number` ?

itsbjoern commented:
> When I compile your version I get this error:
> 
> ```
> src/WebviewManager.ts:122:86 - error TS2345: Argument of type 'number | undefined' is not assignable to parameter of type 'number'.
>   Type 'undefined' is not assignable to type 'number'.
> 
> 122                     await this.handleOpenFile(panel, message.filePath, message.line, message.column);
>                                                                                          ~~~~~~~~~~~~~~
> 
> 
> Found 1 error in src/WebviewManager.ts:122
> ```
> 
> Could you fix this? Maybe we coul use `message.column ?? 0` at the call site, or change `handleOpenFile` to accept `column?: number` ?

Ah, sorry about that. I strongly typed the messages and expected payloads now.
